python/pyspark/sql/session.py

```diff
@@ -414,6 +415,43 @@ def _createFromLocal(self, data, schema):
         data = [schema.toInternal(row) for row in data]
         return self._sc.parallelize(data), schema
 
+    def _createFromPandasWithArrow(self, df, schema):
+        """
+        Create a DataFrame from a given pandas.DataFrame by slicing the into partitions, converting
+        to Arrow data, then reading into the JVM to parallelsize. If a schema is passed in, the
+        data types will be used to coerce the data in Pandas to Arrow conversion.
+        """
+        import os
+        from tempfile import NamedTemporaryFile
+        from pyspark.serializers import ArrowSerializer
+        from pyspark.sql.types import from_arrow_schema, to_arrow_schema
+        import pyarrow as pa
+
+        # Slice the DataFrame into batches
+        step = -(-len(df) // self.sparkContext.defaultParallelism)  # round int up
+        df_slices = (df[start:start + step] for start in xrange(0, len(df), step))
+        arrow_schema = to_arrow_schema(schema) if schema is not None else None
+        batches = [pa.RecordBatch.from_pandas(df_slice, schema=arrow_schema, preserve_index=False)
+                   for df_slice in df_slices]
+
+        # write batches to temp file, read by JVM (borrowed from context.parallelize)
+        tempFile = NamedTemporaryFile(delete=False, dir=self._sc._temp_dir)
+        try:
+            serializer = ArrowSerializer()
+            serializer.dump_stream(batches, tempFile)
+            tempFile.close()
+            readRDDFromFile = self._jvm.PythonRDD.readRDDFromFile
+            jrdd = readRDDFromFile(self._jsc, tempFile.name, len(batches))
+        finally:
+            # readRDDFromFile eagerily reads the file so we can delete right after.
+            os.unlink(tempFile.name)
+
+        # Create the Spark DataFrame, there will be at least 1 batch
+        schema = from_arrow_schema(batches[0].schema)
```

Comment:
`schema = from_arrow_schema(batches[0].schema) if schema is None else schema`?

Reply:
This brings up a good point, the schema should usually be the same, but what should the behavior be if the user specifies the wrong schema?

1. If use the passed in schema and it is wrong (the line in the comment above) then there might be an error later when an operation is performed.  For example, if a string column was specified as DoubleType that would produce:  
"Caused by: java.lang.UnsupportedOperationException at org.apache.spark.sql.execution.vectorized.ArrowColumnVector$ArrowVectorAccessor.getDouble(ArrowColumnVector.java:395)"

2.  We could check the user supplied schema matches the ArrowRecordBatch schema and fail immediately if not equal.  I'm not sure if there might be some cases where you wouldn't want it to fail - like with different integer types..

3.  Always use the schema from the ArrowRecordBatch (how it is currently in this PR)

I'm thinking that (2) is the best since it's the safest and would produce a clear error message.

Reply:
Could we check ahead and fall back to `toPandas` without Arrow? I think I am seeing some differences in some cases, for example:

```python
import pandas as pd
import numpy as np

spark.conf.set("spark.sql.execution.arrow.enabled", "false")
spark.createDataFrame(pd.DataFrame(data={"intcol": [0.8]}), schema="intcol STRING").show()
spark.conf.set("spark.sql.execution.arrow.enabled", "true")
spark.createDataFrame(pd.DataFrame(data={"intcol": [0.8]}), schema="intcol STRING").show()
```

WDYT @BryanCutler and @ueshin ?

Reply:
There might be some differences because without Arrow, Spark just gets the column names from Pandas and infers the data type. This is by design, but I've seen a lot of users get tripped up by it and create JIRAs. With Arrow, it makes it easy to use the schema from Pandas so I would consider this an improvement.

Reply:
But I thought we should reduce the diff between `createDataFrame` with Arrow and `createDataFrame` without Arrow, and match the behaviour first though. To be clear, my suggestion is:

> 2. We could check the user supplied schema matches the ArrowRecordBatch schema and fail immediately if not equal. I'm not sure if there might be some cases where you wouldn't want it to fail - like with different integer types..

but if they are different, fall back to `createDataFrame` without Arrow to reduce the differences between them. I carefully guess `createDataFrame` is stricter?

Reply:
Can we adjust data types if `schema is not None` like we did at `ArrowStreamPandasSerializer.dump_stream`?

Reply:
Btw, do we also need to support schema like `['name', 'age']`, `"int"`(not `StructType`), etc. from doctest.

Reply:
Sorry @HyukjinKwon , I misread your example.  Yes, we can fall back to not use arrow if the user supplied schema doesn't match from arrow - I actually get equal results from your example tho

> can we adjust data types if schema is not None

@ueshin we could do this too, I'm not sure what would be better in this case.. I'll think about it for a bit

Reply:
Hm.. FWIW, I get something like:

```
  File "pyarrow/table.pxi", line 517, in pyarrow.lib.RecordBatch.from_pandas (/.../build/xhochy/pyarrow-macos-wheels/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:28616)
  File "pyarrow/table.pxi", line 336, in pyarrow.lib._dataframe_to_arrays (/.../build/xhochy/pyarrow-macos-wheels/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:26836)
  File "pyarrow/array.pxi", line 1157, in pyarrow.lib.Array.from_pandas (/.../build/xhochy/pyarrow-macos-wheels/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:19825)
  File "pyarrow/error.pxi", line 66, in pyarrow.lib.check_status (/.../build/xhochy/pyarrow-macos-wheels/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:7157)
pyarrow.lib.ArrowNotImplementedError: NotImplemented: string
```

```
>>> import pyarrow
>>> pyarrow.__version__
'0.4.0'
>>> import pandas
>>> pandas.__version__
u'0.20.2'
```

Reply:
Oh pyarrow 0.4.1 is what is installed on Jenkins, so that is what I've been testing against. Maybe try that version?

Reply:
Hmmmm.. still get the same exception with pyarrow 0.4.1 ..